### PR TITLE
[mlir][Vector] Add a rewrite pattern for better low-precision ext(bit…

### PR DESCRIPTION
**Pull request by nicolasvasilache**
…cast) expansion
    
This revision adds a rewrite for sequences of vector `ext(maybe_broadcast(bitcast))`
to use a more efficient sequence of vector operations comprising `shuffle`, `shift` and
`bitwise` ops.
    
The rewrite uses an intermediate bitwidth equal to the licm of
the element types of the source and result types of `bitCastOp`. This
intermediate type may be small or greater than the desired elemental type of
the `ext`, in which case appropriate `ext` or `trunc` operations are inserted.
    
The rewrite fails if the intermediate type is greater than `64` and if the
involved vector types fail to meet basic divisilibity requirements. In other
words, this rewrite does not handle partial vector boundaries and leaves
this part of the heavy-lifting to LLVM.
    
In the future, it may be relevant to give control on the size of the intermediate type.
For now, it is empirically determined that taking `64` result in much better assembly
being produced when piping through `llvm-mca`.

### CODE REVIEW

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 1]
```diff
@@ -155,6 +170,256 @@ struct ConvertVectorTransferRead final
 };
 } // end anonymous namespace
 
+//===----------------------------------------------------------------------===//
+// RewriteExtOfBitCast
+//===----------------------------------------------------------------------===//
+
+/// Create a vector of bit masks: `idx .. idx + step - 1` and broadcast it
```
Comment (ftynse) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:177]: It's unclear what `idx` means in this comment.

A second example could also be helpful to understand the logic generalization.

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 2]
```diff
@@ -155,6 +170,256 @@ struct ConvertVectorTransferRead final
...
+/// Create a vector of bit shuffles: `numOccurrences * idx` and broadcast it
+/// `bitwidth/step` times.
+/// `step` must divide `bitwidth` evenly.
+/// Example: (4, 2, 3) -> [0x0, 0x1, 0x0, 0x1, 0x0, 0x1].
+static SmallVector<int64_t>
+computeExtOfBitCastShuffles(MLIRContext *ctx, int64_t bitwidth, int64_t step,
+                            int64_t numOccurrences) {
+  assert(bitwidth % step == 0 && "step must divide bitwidth evenly");
+  SmallVector<int64_t> shuffles;
+  int64_t n = floorDiv(bitwidth, step);
```
Comment (ftynse) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:234]: Why floorDiv if `step` is known to evenly divide `bitwidth` given the assert above?

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 3]
```diff
@@ -155,6 +170,256 @@ struct ConvertVectorTransferRead final
...
+  }
+  return 0;
+}
+
+FailureOr<Value>
+mlir::vector::rewriteExtOfBitCast(RewriterBase &rewriter, Operation *extOp,
+                                  vector::BitCastOp bitCastOp,
+                                  vector::BroadcastOp maybeBroadcastOp) {
+  assert(
+      (llvm::isa<arith::ExtSIOp>(extOp) || llvm::isa<arith::ExtUIOp>(extOp)) &&
```
Comment (ftynse) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:265]: Nit: no need to prefix with `llvm::`, this name is re-exported.
Also nit: `isa` is a variadic template, `isa<arith::ExtSIOp, ExtUIOp>()` should work.

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 4]
```diff
@@ -155,6 +170,256 @@ struct ConvertVectorTransferRead final
...
+  Location loc = extOp->getLoc();
+  MLIRContext *ctx = extOp->getContext();
+
+  VectorType interimVectorType =
+      VectorType::Builder(sourceVectorType)
+          .setDim(sourceVectorType.getRank() - 1, interimMostMinorDim)
+          .setElementType(IntegerType::get(ctx, interimBitWidth));
+  LDBG("interimVectorType: " << interimVectorType);
+
+  IntegerType interimIntType = IntegerType::get(ctx, interimBitWidth);
```
Comment (ftynse) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:304]: Nit: put above `interimVectorType` and use in its construction.

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 5]
```diff
@@ -155,6 +170,256 @@ struct ConvertVectorTransferRead final
...
+  VectorType resultType = cast<VectorType>(extOp->getResultTypes().front());
+  Type resultElementType = getElementTypeOrSelf(resultType);
+  SmallVector<Attribute> shifts = computeExtOfBitCastShifts(
+      ctx, interimBitWidth, targetBitWidth, interimMostMinorDim);
+  auto shiftConstantOp = rewriter.create<arith::ConstantOp>(
+      loc, DenseElementsAttr::get(vt, shifts));
+  LDBG("shiftConstant: " << shiftConstantOp);
+  Value newResult =
+      TypeSwitch<Operation *, Value>(extOp)
+          .template Case<arith::ExtSIOp>([&](arith::ExtSIOp op) {
```
Comment (ftynse) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:341]: Nit: I don't think `.template` is needed here.

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 6]
```diff
@@ -155,6 +170,256 @@ struct ConvertVectorTransferRead final
...
+            auto vt = shifted.getType().cast<VectorType>();
+            VectorType extVt =
+                VectorType::Builder(vt).setElementType(resultElementType);
+            Operation *res =
+                (resultElementType.getIntOrFloatBitWidth() > interimBitWidth)
+                    ? rewriter.create<arith::ExtSIOp>(loc, extVt, shifted)
+                    : rewriter.create<arith::TruncIOp>(loc, extVt, shifted);
+            return res->getResult(0);
+          })
+          .template Case<arith::ExtUIOp>([&](arith::ExtUIOp op) {
```
Comment (ftynse) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:353]: If you wanted to go full template here, you could do something like

```c++
template <typename ExtOp,
    typename ShiftOp = std::conditional_t<std::is_same_v<ExtOp, arith::ExtUIOp>, arith::ShRUIOp, arith::ShRSIOp>>
static FailureOr<Value>
rewriteExtOfBitCastImpl (RewriterBase &rewriter, ExtOp op, ...) {
  // ...
  Value shifted = rewriter.template create<ShiftOp>(...);
  // ...
}
```

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 7]
```diff
@@ -155,6 +170,256 @@ struct ConvertVectorTransferRead final
...
+                    ? rewriter.create<arith::ExtUIOp>(loc, extVt, shifted)
+                    : rewriter.create<arith::TruncIOp>(loc, extVt, shifted);
+            return res->getResult(0);
+          })
+          .Default([&](Operation *op) {
+            llvm_unreachable("unexpected op type");
+            return nullptr;
+          });
+
+  if (maybeBroadcastOp) {
```
Comment (ftynse) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:370]: Nit: return early instead.

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 8]
```diff
@@ -155,6 +170,256 @@ struct ConvertVectorTransferRead final
...
+      : OpRewritePattern<ExtOpType>(context, benefit) {}
+
+  LogicalResult matchAndRewrite(ExtOpType extOp,
+                                PatternRewriter &rewriter) const override {
+    VectorType resultTy = dyn_cast<VectorType>(extOp.getType());
+    if (!resultTy)
+      return rewriter.notifyMatchFailure(extOp, "not a vector type");
+
+    int64_t elementalBitWidth = resultTy.getElementTypeBitWidth();
+    if (elementalBitWidth & (elementalBitWidth - 1)) {
```
Comment (ftynse) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:393]: Nit: I'd rather use `llvm::isPowerOf2_64`

File: mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+!mtt = memref<1234x!final_element_t>
+
+// CHECK-LABEL: @f1(
+//  CHECK-SAME: %[[A:[a-z0-9]+]]: memref<1234xi8>
+//  CHECK-SAME: %[[IDX:[a-z0-9]+]]: index
+//  CHECK-SAME: %[[B:[a-z0-9]+]]: memref<1234xf32>
+func.func @f1(%m: !mst, %idx : index, %mf: !mtt) {
+
+//              CHECK: %[[MASK:.*]] = arith.constant dense<[
+// CHECK-SAME-COUNT-6: 7, 56, 448, 3584, 28672, 229376, 1835008, 14680064, 117440512, 939524096, 7516192768, 60129542144, 481036337152, 3848290697216, 30786325577728, -35184372088832
```
Comment (ftynse) [inline mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir:32]: A comment with the hex version of this would be helpful.

**Review by ftynse — APPROVED** [review]

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 10]
```diff
@@ -155,6 +170,256 @@ struct ConvertVectorTransferRead final
...
+  SmallVector<Attribute> shifts;
+  for (int64_t idx = 0; idx < numOccurrences; ++idx)
+    llvm::append_range(shifts, tmpShifts);
+  return shifts;
+}
+
+/// Create a vector of bit shuffles: `numOccurrences * idx` and broadcast it
+/// `bitwidth/step` times.
+/// `step` must divide `bitwidth` evenly.
+/// Example: (4, 2, 3) -> [0x0, 0x1, 0x0, 0x1, 0x0, 0x1].
```
Comment (qcolombet) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:228]: I have a hard time reconciling the comment and the example.

For instance if we broadcast `bitwidth/step` times, we should only have 2 reps, not 3.

File: mlir/test/Dialect/LLVM/transform-e2e.mlir [hunk 11]
```diff
@@ -29,33 +29,12 @@ transform.sequence failures(propagate) {
   // lowering TD macros.
   transform.apply_patterns to %f {
     transform.apply_patterns.vector.lower_contraction lowering_strategy = "outerproduct"
-  } : !transform.any_op
-
-  transform.apply_patterns to %f {
```
Comment (qcolombet) [inline mlir/test/Dialect/LLVM/transform-e2e.mlir:34]: These changes are unrelated, right?

**Review by qcolombet — APPROVED** [review]
LGTM.
+1 on @ftynse comments plus a few of my own :).

**Review by stellaraccident — COMMENTED** [review]
Looks like the review is in good hands, so I'll just +1 having this. Thank you!

**Comment by nicolasvasilache** [comment]
I need to rework this to reuse the same implementation as https://github.com/llvm/llvm-project/pull/66387 which is both more general and more performant overall.

**Comment by nicolasvasilache** [comment]
This is now reimplemented in terms of #66387, closing this and starting another PR
